## /epic-review — PR #9739 (for Epic #9724 / sub-issue #9725)

**Reviewer**: Tech Lead (via `/epic-review`)

### 1. Requirements (Definition of Done from #9725)

| # | Requirement | Status | Notes |
|---|-------------|--------|-------|
| 1 | `createZeroRun` body uses `after(() => dispatchZeroRun(...).catch(...))` | ✅ | `zero-run-service.ts:636-643`, error logged + still persisted via `markRunFailed` inside `dispatchZeroRun` |
| 2 | `createZeroRunRecord` and `dispatchZeroRun` un-exported | ✅ | Both now `async function` (no `export`) — verified via grep |
| 3 | `/api/zero/runs/route.ts`: `after` import removed, single `createZeroRun(...)` call | ✅ | No `after` in imports; line 124 awaits `createZeroRun` directly |
| 4 | `/api/zero/chat/messages/route.ts`: dispatch `after()` removed, realtime-signal `after()` preserved | ✅ | Dispatch is now single `await createZeroRun(...)` at line 219; realtime-signal `after(async () => {...})` at line 249 preserved as planned |
| 5 | `zero-run-service.test.ts`: deferred-dispatch witness test | ✅ | New test asserts runner job entry absent before `flushAfter()`, present after |
| 6 | `build-zero-context.test.ts`: `flushAfter()` added before `findTestRunnerJobEntry` lookups | ✅ | 10 lookups updated |

### 2. Code Review

| Dimension | Status | Findings |
|-----------|--------|----------|
| Correctness | ✅ | Phase 1/2 split preserved. Dispatch correctly guarded by `if (result.record)` — enqueued runs (concurrency-limited) are drained by the queue worker later, not dispatched. |
| Project Standards | ✅ | No `any`, no `eslint-disable`. Error handling is targeted (Phase 2 failures caught and persisted via `markRunFailed`). |
| Consistency | ✅ | Follows the existing `after()` pattern from Epic #9682. `flushAfter()` helper update (iterative drain) is a clean generalization. |
| Security | ✅ | No change to auth/validation; contract at route boundary unchanged. |
| Performance | ✅ | Every caller now gets the ~1–1.5s response-time improvement automatically. |

### 3. Architectural Alignment

| Check | Status | Notes |
|-------|--------|-------|
| `createZeroRun` is the only exported run-creation API | ✅ | `createZeroRunRecord` and `dispatchZeroRun` now module-internal |
| `after()` primitive hidden from call sites | ✅ | Routes no longer import `after` for dispatch; realtime-signal `after()` in chat/messages is a separate concern |
| No explicit `createZeroRunRecord + after()` dance at call sites | ✅ | Both optimized routes collapsed |

### 4. Issues Found

#### Must Fix (blocking)

1. **`zero-run-service.ts:610` — `interface CreateZeroRunResult` is not exported**, but `createZeroRun()` returns `Promise<CreateZeroRunResult>` and is consumed by exported voice-chat dispatch functions whose inferred return types now include this type.

   CI `lint-types` failure (TS4058):
   ```
   src/lib/zero/voice-chat/preparation-service.ts(164,23): error TS4058: Return type of exported function has or is using name 'CreateZeroRunResult' ... but cannot be named.
   src/lib/zero/voice-chat/session-service.ts(143,23): error TS4058: ...
   src/lib/zero/voice-chat/session-service.ts(241,23): error TS4058: ...
   ```

   **Fix**: Change `interface CreateZeroRunResult {` → `export interface CreateZeroRunResult {` at line 610.

### Verdict

**NEEDS WORK** — one-line fix. Core refactor is correct and well-executed; the only blocker is an unexported interface tripping TS4058 on downstream exported consumers.

--- a/turbo/apps/web/src/lib/zero/slack-org/handlers/run-agent.ts
+++ b/turbo/apps/web/src/lib/zero/slack-org/handlers/run-agent.ts
@@ -27,7 +27,7 @@ interface RunAgentParams {
 }
 
 interface RunAgentResult {
-  status: "dispatched" | "queued" | "failed";
+  status: "accepted" | "queued" | "failed";
   response?: string;
   runId: string | undefined;
   errorCode?: string;
@@ -91,7 +91,7 @@ export async function runAgentForSlackOrg(
       ],
     });
 
-    const status = result.status === "queued" ? "queued" : "dispatched";
+    const status = result.status === "queued" ? "queued" : "accepted";
     log.debug(`Run ${result.runId} ${status} for Slack org agent ${agentName}`);
 
     return { status, runId: result.runId };

--- a/turbo/apps/web/src/lib/zero/telegram/handlers/run-agent.ts
+++ b/turbo/apps/web/src/lib/zero/telegram/handlers/run-agent.ts
@@ -21,7 +21,7 @@ interface RunAgentParams {
 }
 
 interface RunAgentResult {
-  status: "dispatched" | "queued" | "failed";
+  status: "accepted" | "queued" | "failed";
   response?: string;
   runId: string | undefined;
 }
@@ -68,7 +68,7 @@ export async function runAgentForTelegram(
       ],
     });
 
-    const status = result.status === "queued" ? "queued" : "dispatched";
+    const status = result.status === "queued" ? "queued" : "accepted";
     log.debug(`Run ${result.runId} ${status} for Telegram agent ${agentName}`);
 
     return {

--- a/turbo/apps/web/src/lib/zero/voice-chat/preparation-service.ts
+++ b/turbo/apps/web/src/lib/zero/voice-chat/preparation-service.ts
@@ -1,6 +1,6 @@
 import { eq, and, gt, lt, desc, isNull } from "drizzle-orm";
 import { voiceChatPreparations } from "../../../db/schema/voice-chat";
-import { createZeroRun } from "../zero-run-service";
+import { createZeroRun, type CreateZeroRunResult } from "../zero-run-service";
 import {
   buildVoiceChatPrepareOnlyPrompt,
   buildVoiceChatMeetingPreparePrompt,
@@ -166,7 +166,7 @@ export async function dispatchPreparationRun(
   userId: string,
   agentId: string,
   options?: { mode?: "chat" | "meeting"; prompt?: string },
-) {
+): Promise<CreateZeroRunResult> {
   const db = globalThis.services.db;
   const meetingPrompt =
     options?.mode === "meeting" ? options.prompt : undefined;

--- a/turbo/apps/web/src/lib/zero/voice-chat/session-service.ts
+++ b/turbo/apps/web/src/lib/zero/voice-chat/session-service.ts
@@ -3,7 +3,7 @@ import {
   voiceChatSessions,
   voiceChatEvents,
 } from "../../../db/schema/voice-chat";
-import { createZeroRun } from "../zero-run-service";
+import { createZeroRun, type CreateZeroRunResult } from "../zero-run-service";
 import { cancelRun } from "../zero-run-cancel";
 import {
   buildVoiceChatQuickPrepPrompt,
@@ -146,7 +146,7 @@ export async function dispatchSlowBrain(
   userId: string,
   agentId: string,
   options?: { mode?: "chat" | "meeting"; prompt?: string },
-) {
+): Promise<CreateZeroRunResult> {
   const db = globalThis.services.db;
   const meetingPrompt =
     options?.mode === "meeting" ? options.prompt : undefined;
@@ -242,7 +242,7 @@ export async function dispatchObservationSlowBrain(session: {
   id: string;
   userId: string;
   agentId: string;
-}) {
+}): Promise<CreateZeroRunResult> {
   const db = globalThis.services.db;
   const appendSystemPrompt = buildVoiceChatObservationOnlyPrompt(session.id);
   const prompt = `You are Zero's slow-brain for voice-chat session ${session.id}. Preparation is complete. Start observing the conversation.`;

--- a/turbo/apps/web/src/lib/zero/zero-run-service.ts
+++ b/turbo/apps/web/src/lib/zero/zero-run-service.ts
@@ -606,8 +606,12 @@ async function dispatchZeroRun(
  * Only fields populated by Phase 1 (pre-flight + INSERT) are exposed; Phase 2
  * (tokens, context, dispatch) runs deferred inside after() so its outputs
  * (sandboxId, final dispatched status) are not available at return time.
+ *
+ * `status` reflects Phase 1 state only — it is always `"pending"` (record
+ * inserted, dispatch scheduled via after()) or `"queued"` (concurrency limit,
+ * will be dispatched by the queue worker). It is NEVER a post-dispatch status.
  */
-interface CreateZeroRunResult {
+export interface CreateZeroRunResult {
   runId: string;
   status: RunStatus;
   createdAt: Date;

